src/v2/Artsy/Relay/middleware/cache/Cache.ts

```diff
@@ -0,0 +1,147 @@
+import createLogger from "v2/Utils/logger"
+import RelayQueryResponseCache from "relay-runtime/lib/network/RelayQueryResponseCache"
+import { isServer } from "lib/environment"
+import { RetryStrategyOptions } from "redis"
+
+import {
+  GraphQLResponse,
+  QueryResponseCache,
+  CacheConfig as RelayCacheConfig,
+  Variables as RelayVariables,
+} from "relay-runtime"
+
+const logger = createLogger("v2/Artsy/middleware/cache/Cache")
+
+export interface CacheConfig {
+  size: number
+  ttl: number
+  disableServerSideCache?: boolean
+}
+
+export class Cache {
+  cacheConfig: CacheConfig
+  enableServerSideCache: boolean
+  relayCache: RelayQueryResponseCache
+
+  redisCache: {
+    del: (key: string) => Promise<void>
+    expire: (key: string, ttl: number) => Promise<void>
+    flushall: () => Promise<void>
+    get: (key: string) => Promise<string>
+    set: (key: string, value: string) => Promise<void>
+  }
+
+  constructor(cacheConfig: CacheConfig) {
+    this.cacheConfig = cacheConfig
+    this.enableServerSideCache =
+      isServer && !this.cacheConfig.disableServerSideCache
+    this.initRelayCache()
+    this.initRedisCache()
+  }
+
+  initRelayCache() {
+    this.relayCache = new QueryResponseCache(this.cacheConfig)
+  }
+
+  initRedisCache() {
+    if (!this.enableServerSideCache) {
+      return
+    }
+
+    const { promisify } = require("util")
+    const redis = require("redis")
+
+    const client = redis.createClient({
+      url: process.env.OPENREDIS_URL,
+      retry_strategy: this.handleRedisError,
+    })
+
+    this.redisCache = {
+      del: promisify(client.del).bind(client),
+      expire: promisify(client.expire).bind(client),
+      flushall: promisify(client.flushall).bind(client),
+      get: promisify(client.get).bind(client),
+      set: promisify(client.set).bind(client),
+    }
+  }
+
+  handleRedisError(retryOptions: RetryStrategyOptions) {
+    const TIMEOUT = Number(process.env.PAGE_CACHE_RETRIEVAL_TIMEOUT_MS)
+    const MAX_RETRIES = 10
```

Comment:
> I'm not sure of the semantics of this retry_strategy argument. Are these standard or copied from somewhere?

Yup, copied [from the docs](https://github.com/NodeRedis/node-redis) and seemed to make sense as defaults. 

> but the problems we've encountered have surfaced when individual fetches can't reach the redis host

This is a generic error handler and that case has been covered [here](https://github.com/artsy/force/pull/5684/files#diff-628b8502c71aed0fa389099db155b0bcR78-R80). There are a number of different cases we can evaluate. I think this might need to be tuned / explored a bit more but it seems like we should be going through formal channels, even if we'd want to fail immediately (no retries).

Reply:
I'm not sure what you mean by formal channels so I'll just narrow down my question: how would this behave when the cache starts taking a long time to respond? We've encountered this in the past and it's caused avoidable incidents.

Reply:
I agree with @joeyAghion here. It feels strange to implement exponential backoff for caching. If there is a cache miss or error I would expect it to immediately continue and fallback to the original logic.

If we do keep this retry strategy I would prefer that one strategy be used: either a timeout or exponential backoff. Currently the two strategies are at conflict with one another.

With `MAX_RETRIES` of 10, the total that is allowed before giving up is 5500ms, however the `PAGE_CACHE_RETRIEVAL_TIMEOUT_MS` is set to 400ms meaning we will only ever achieve 3 retries (100ms + 200ms + 300ms) before hitting timeout limit.

Reply:
Yea, I will double check, but I recall setting my redis endpoint to an intentionally delayed endpoint from http://slowwly.robertomurray.co.uk/ , and testing that the various cache client timeout options didn't seem to shield the app from awaiting on a slow endpoint, and not responding to the request within the configured timeout.

So overall I would vote for dropping those in favor of our known working `setTimeout` wrapped way (which has the overhead of adding a promise and timer to things). I recall testing that with a slow endpoint and it worked as desired.

Reply:
(It could be I'm thinking of Metaphysics and memcache, in which case it's definitely worth double-checking using the built-in client handling).

Reply:
Cool thanks ya'll, I'll make that change 👍